File: Solutions/Increasing-Order-Search-Tree.py
# Definition for a binary tree node.
from typing import List

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def increasingBST(self, root: TreeNode) -> TreeNode:
        ans = []
        if (root): self.recurve(root, ans)
        
        new_root = TreeNode(ans[0])
        cur_root = new_root
        for i in range(1, len(ans)):
            cur_root.right = TreeNode(ans[i])
            cur_root = cur_root.right
        return new_root
        
    def recurve(self, root: TreeNode, ans: List) -> None:
        if (root.left): self.recurve(root.left, ans)
        ans.append(root.val)
        if (root.right): self.recurve(root.right, ans)

T1 = TreeNode(1)
T3 = TreeNode(3)
T2 = TreeNode(2, T1, T3)
S = Solution()
S.increasingBST(T2)

        

 

File: Solutions/Increasing-Order-Search-Tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
  public:
    vector<TreeNode *> v;
    void _increasingBST(TreeNode *root)
    {
        if (root == NULL)
            return;

        _increasingBST(root->left);
        v.push_back(root);
        _increasingBST(root->right);
        return;
    };
    TreeNode *increasingBST(TreeNode *root)
    {
        _increasingBST(root);
        int n = v.size();
        for (int i = 0; i < n - 1; i++)
        {
            v[i]->left = nullptr;
            v[i]->right = v[i + 1];
        };
        v[n - 1]->left = v[n - 1]->right = nullptr;
        return v[0];
    }
};